### PR TITLE
Experimental mixed combinations

### PR DESCRIPTION
**Pull request by LukasZahradnik**
Works only for + (sum), * (elprod), and @ (prod) for now

Added support for references:
```
R.a(V.X) <= (R.b(V.X) + R.c(1) * (d := R.d(2)) * d + d * d)
```
This works! There will be actually only three literals in the body

### CODE REVIEW

File: neuralogic/core/constructs/rule.py [hunk 1]
```diff
@@ -86,12 +90,12 @@ def __init__(self, head, body):
...
-                continue
-            for term in body_atom.terms:
-                if term is Ellipsis:
-                    return True
+        # for body_atom in self.body:
+        #     if not body_atom.predicate.special or body_atom.predicate.name != "alldiff":
+        #         continue
+        #     for term in body_atom.terms:
+        #         if term is Ellipsis:
+        #             return True
```
Comment (LukasZahradnik) [inline neuralogic/core/constructs/rule.py:98]: This should work but will cause an error once we replace the templated body literal

File: neuralogic/core/constructs/function/__init__.py [hunk 2]
```diff
@@ -29,4 +32,78 @@
 Aggregation.SOFTMAX = Softmax("SOFTMAX")
 
 
-__all__ = ["Transformation", "Combination", "Aggregation", "Function"]
+class CombinationWrap:
+    __slots__ = "left", "right", "combination"
```
Comment (LukasZahradnik) [inline neuralogic/core/constructs/function/__init__.py:36]: Currently only works for binary operators. Consider extending this to support combinations on n elements.

File: neuralogic/core/constructs/function/__init__.py [hunk 3]
```diff
@@ -29,4 +32,78 @@
...
+            for a in self.right:
+                yield a
+        if not isinstance(self.right, CombinationWrap):
+            yield self.right
+
+    def to_combination(self) -> Combination:
+        combination_graph, _ = self._get_combination_node()
+        return MixedCombination(name=self.to_str(), combination_graph=combination_graph)
+
+    def _get_combination_node(self, input_counter: int = 0):
```
Comment (LukasZahradnik) [inline neuralogic/core/constructs/function/__init__.py:79]: This whole implementation kinda assumes that the whole body is a mixed combination construct.

RuleBody should:
- Check if there is a CombinationWrap in its body
- Make the whole body into CombinationWrap by using either Combination provided in metadata OR the default one.
- This has to happen before building - after we have metadata ready

File: neuralogic/core/constructs/function/function.py [hunk 4]
```diff
@@ -83,6 +86,16 @@ class Combination(Function):
...
 
+    def get(self):
+        name = self.name.capitalize()
+
+        if name in ("Sum", "Max", "Min", "Avg", "Count"):
+            return jpype.JClass(f"cz.cvut.fel.ida.algebra.functions.aggregation.{name}")()
+        if name == "Elproduct":
+            name = "ElementProduct"
+
+        return jpype.JClass(f"cz.cvut.fel.ida.algebra.functions.combination.{name}")()
```
Comment (LukasZahradnik) [inline neuralogic/core/constructs/function/function.py:97]: Will this work with parametrized combinations? E.g., concat?

Comment (LukasZahradnik) [reply neuralogic/core/constructs/function/function.py:97]: It wont - combinations are now defined as unary functions (`F.concat(entity, axis)`) 💀💀

File: neuralogic/core/constructs/function/__init__.py [hunk 5]
```diff
@@ -29,4 +32,78 @@
 Aggregation.SOFTMAX = Softmax("SOFTMAX")
 
 
-__all__ = ["Transformation", "Combination", "Aggregation", "Function"]
+class CombinationWrap:
```
Comment (LukasZahradnik) [inline neuralogic/core/constructs/function/__init__.py:35]: This whole thing could support also transformation functions? Might be nice

File: neuralogic/core/constructs/function/__init__.py [hunk 6]
```diff
@@ -29,4 +32,78 @@
...
+
+        if isinstance(self.right, CombinationWrap):
+            right_node, input_counter = self.right._get_combination_node(input_counter)
+        else:
+            right_index = input_counter
+            input_counter += 1
+
+        class_name = "cz.cvut.fel.ida.algebra.functions.combination.MixedCombination.MixedCombinationNode"
+
+        return (
```
Comment (LukasZahradnik) [inline neuralogic/core/constructs/function/__init__.py:100]: Almost feel like there should be some simplification to combine all combinations into one if they are the same and the order is not changes.

That is, instead of sum(a, sum(b, sum(c, d))) have sum(a, b, c, d)

Comment (LukasZahradnik) [reply neuralogic/core/constructs/function/__init__.py:100]: Some combinations might not be permutation invariant - don't really want to deal with that, so do the simplification only when literals are in the original sequence?

File: neuralogic/core/constructs/rule.py [hunk 7]
```diff
@@ -54,7 +55,10 @@ def __init__(self, head, body):
         if not isinstance(body, Iterable):
             body = [body]
 
-        self.body = list(body)
+        self.body = body
```
Comment (LukasZahradnik) [inline neuralogic/core/constructs/rule.py:58]: There is a proposed syntax for lambdas, e.g.,:
```
R.a <= (b := R.b, c := R.c) >> F.tanh(b + c)
```

This will allow a lot of messy scenarios:
```
R.a <= (F.relu(b := R.b) * (c := R.c)) >> F.tanh(b + c)
```

Comment (LukasZahradnik) [reply neuralogic/core/constructs/rule.py:58]: This need some more thinking. What if we have the following:

```
F.relu(R.a) <= (R.b + R.c) | [F.avg]
```
Which would correspond to:

```
(R.a <= (R.b, R.c)) | [F.avg, F.sum, F.relu]
```

Comment (LukasZahradnik) [reply neuralogic/core/constructs/rule.py:58]: There will always be just one aggregation function, as it aggregates single values from all rules into one. Transformation on head could be stacked (e.g., multiple `tanh(relu(R.a))`, the same way could be stacked functions for relations in the body.

Comment (LukasZahradnik) [reply neuralogic/core/constructs/rule.py:58]: The issue is readability - it's mixing computation with logic - but we already do that with soft negation hack (activation functions in the body is already possible)

Comment (LukasZahradnik) [reply neuralogic/core/constructs/rule.py:58]: We also cannot do the following `F.tanh(b + c * c)` - that might be possible with some hacking?

```
F.relu(R.a) <= (R.b + (c  := R.c) * c) | [F.avg]
```

File: neuralogic/core/constructs/function/__init__.py (hunk 2 above)
Comment (LukasZahradnik) [reply neuralogic/core/constructs/function/__init__.py:38]: Extending this to n elements would mean supporting transformations out-of-the-box (???)
